sapicore/engine/network/__init__.py

```diff
@@ -0,0 +1,400 @@
+                integrated_data = self.summation([synapse.output for synapse in incoming_synapses]).to(self.device)
+
+            else:
+                # if this is a root node, treat its stream from the data loader as inbound synaptic input.
+                external = [data[self.roots.index(ensemble_ref.identifier)]] if isinstance(data, list) else [data]
+                feedback = [synapse.output for synapse in incoming_synapses]
+
+                integrated_data = self.summation(external + feedback)
+
+            # forward current ensemble.
+            ensemble_ref(integrated_data)
+
+            # use current ensemble's analog or spike data to forward its outgoing synapses.
+            if hasattr(ensemble_ref, "spiked"):
+                [synapse(ensemble_ref.spiked.float()) for synapse in outgoing_synapses]
+
+            elif hasattr(ensemble_ref, "voltage"):
+                [synapse(ensemble_ref.voltage) for synapse in outgoing_synapses]
+
+        self.simulation_step += 1
+
+        # build and pass a state dictionary for use by calling module.
+        state = {"ensembles": {}, "synapses": {}}
+        for comp in self.get_ensembles():
+            state["ensembles"][comp.identifier] = comp.loggable_state()
+
+        for comp in self.get_synapses():
+            state["synapses"][comp.identifier] = comp.loggable_state()
+
+        return state
+
+    def _in_edges(self, node: str) -> list[Synapse]:
+        """Returns list of synapse objects going into the ensemble `node`."""
+        return [ref.get("reference") for _, _, ref in list(self.graph.in_edges(node, data=True))]
+
+    def _out_edges(self, node: str) -> list[Synapse]:
+        """Returns list of synapse objects fanning out of the ensemble `node`."""
+        return [ref.get("reference") for _, _, ref in list(self.graph.out_edges(node, data=True))]
+
+    def _find_roots(self):
+        """Identifies nodes that are exposed to external input. Those are used as starting points for the multi-BFS
+        forward sweep, whose order also gets updated every time this method is called.
+
+        Warning
+        -------
+        Roots are updated on network initialization and on every application of :meth:`add_ensemble` and
+        :meth:`add_synapse`. If adding a synapse results in a rootless recurrent network, all nodes will have
+        in-degree > 0. In such cases, the method will mark the first ensemble added `self.get_ensembles()[0]`
+        as the root and direct external input to it. This behavior can be modified by setting `self.roots`
+        directly after building the network.
+
+        """
+        updated_roots = [node for node, degree in self.graph.in_degree() if degree == 0]
+        if updated_roots:
+            self.roots = updated_roots
+
+        elif self.get_ensembles():
+            # if the network became rootless after adding a synapse, default to the first ensemble as the root.
+            self.roots = [self.get_ensembles()[0].identifier]
```

Comment:
Rootless could also be two ensembles connected bidirectionally, so neither has in degree 0 (the criterion used to mark something as root in the default implementation). Hypothetically, if one of those is an oscillator, you also don't need external input. That's what prompted this.

I think it would be inconsistent to detect ensembles with only a self-synapse (indeg=0) as roots. Basically, the contract is that by default, if it doesn't have anything going into it it'll be assumed a root. If you have multiple roots, you can input multiple data streams and they will be assigned to those roots (i-th to i-ith). But if you want something else you'd have to specify it. It's completely possible that some experiments might require feeding external input to a center node.

Reply:
This is what I've converged on:

1. Manual `.roots` selection by user should supersede everything, even if that happens at instantiation. I'll make the necessary modifications.

2. If user didn't pick, nodes with indeg==0 are marked (this is stated in the docstring/contract).

3. If all indeg > 0, not specifying any roots would cause issues if the user tries to .fit(data). Picking the first ensemble in the spec is a nice convention--e.g., if you just want to test a PING compartment (circular network), you know that external input will be passed to the first ensemble you specified unless you set `roots` yourself. Better than throwing an error.

Reply:
>This is what I've converged on:

Sounds great. However, please separate them into three different properties/functions so that you can always introspect to see which roots user specified, which have indeg==0,  and the first ensemble. And then have a final `@property` with only a getter that goes through these other three in the order given and returns the first that is not empty (or none if there's no ensembles!?).

Reply:
Already implemented with one `roots` attribute a `root_lock` for ensuring user-setting primacy and for optimizing `build()`. See if you like it in rev IV. Guaranteeing this behavior by construction makes more sense to me than using a complex getter to juggle all scenarios. When you're done instantiating/adding to the network, your (one and only) `roots` list is stable.